perl/man/man3/metaclass.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.31)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "metaclass 3"
.TH metaclass 3 "2015-08-27" "perl v5.22.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
metaclass \- a pragma for installing and using Class::MOP metaclasses
.SH "VERSION"
.IX Header "VERSION"
version 2.1603
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  package MyClass;
\&
\&  # use Class::MOP::Class
\&  use metaclass;
\&
\&  # ... or use a custom metaclass
\&  use metaclass \*(AqMyMetaClass\*(Aq;
\&
\&  # ... or use a custom metaclass
\&  # and custom attribute and method
\&  # metaclasses
\&  use metaclass \*(AqMyMetaClass\*(Aq => (
\&      \*(Aqattribute_metaclass\*(Aq => \*(AqMyAttributeMetaClass\*(Aq,
\&      \*(Aqmethod_metaclass\*(Aq    => \*(AqMyMethodMetaClass\*(Aq,
\&  );
\&
\&  # ... or just specify custom attribute
\&  # and method classes, and Class::MOP::Class
\&  # is the assumed metaclass
\&  use metaclass (
\&      \*(Aqattribute_metaclass\*(Aq => \*(AqMyAttributeMetaClass\*(Aq,
\&      \*(Aqmethod_metaclass\*(Aq    => \*(AqMyMethodMetaClass\*(Aq,
\&  );
\&
\&  # if we\*(Aqd rather not install a \*(Aqmeta\*(Aq method, we can do this
\&  use metaclass meta_name => undef;
\&  # or if we\*(Aqd like it to have a different name,
\&  use metaclass meta_name => \*(Aqmy_meta\*(Aq;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This is a pragma to make it easier to use a specific metaclass
and a set of custom attribute and method metaclasses. It also
installs a \f(CW\*(C`meta\*(C'\fR method to your class as well, unless \f(CW\*(C`undef\*(C'\fR
is passed to the \f(CW\*(C`meta_name\*(C'\fR option.
.PP
Note that if you are using Moose, you most likely do \fBnot\fR want
to be using this \- look into Moose::Util::MetaRole instead.
.SH "AUTHORS"
.IX Header "AUTHORS"
.IP "\(bu" 4
Stevan Little <stevan.little@iinteractive.com>
.IP "\(bu" 4
Dave Rolsky <autarch@urth.org>
.IP "\(bu" 4
Jesse Luehrs <doy@tozt.net>
.IP "\(bu" 4
Shawn M Moore <code@sartak.org>
.IP "\(bu" 4
יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>
.IP "\(bu" 4
Karen Etheridge <ether@cpan.org>
.IP "\(bu" 4
Florian Ragwitz <rafl@debian.org>
.IP "\(bu" 4
Hans Dieter Pearcey <hdp@weftsoar.net>
.IP "\(bu" 4
Chris Prather <chris@prather.org>
.IP "\(bu" 4
Matt S Trout <mst@shadowcat.co.uk>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
This software is copyright (c) 2006 by Infinity Interactive, Inc..
.PP
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


CHANGES.txt
ActivePerl 5.22 Change Log
    For the latest information on ActivePerl, please see:

        http://www.ActiveState.com/ActivePerl/

  Build 2201 January, 2016
    Build 2201 is based on Perl 5.22.1 plus additional selected changes.

    *Bug Fixes and Changes since build 2200*

    *  The following security vulnerabilities have been addressed:

       OpenSSL - multiple fixes
           OpenSSL has been upgraded to 1.0.2e to address several moderate
           issues. Please see
           https://www.openssl.org/news/secadv/20151203.txt for full
           details.

           CVE-2015-3193: BN_mod_exp may produce incorrect results on
           x86_64

           CVE-2015-3194: Certificate verify crash with missing PSS
           parameter

           CVE-2015-3195: X509_ATTRIBUTE memory leak

           CVE-2015-3196: Race condition handling PSK identify hint

           CVE-2015-1794: Anon DH ServerKeyExchange with 0 p parameter

       CVE-2015-8607
           Beginning in PathTools 3.47 and/or perl 5.20.0, the
           File::Spec::canonpath() routine returned untained strings even
           if passed tainted input. This defect undermines the guarantee
           of taint propagation, which is sometimes used to ensure that
           unvalidated user input does not reach sensitive code.

       CVE-2015-8608
           Perl 5 on Windows suffers from two out-of-bounds read and
           multiple small buffer over-read vulnerabilities in the
           VDir::MapPathA and VDir::MapPathW functions that could
           potentially be exploited to achieve arbitrary code execution.

           These defects have been present since
           perl-5.005_02-2346-g7766f13, circa 1999, and are still present
           in the latest releases of Perl.

    *  Significant changes that have occurred in the Perl 5.22.1 release
       are documented in perl5221delta.

    *  Most bundled modules have been updated to their latest released
       version from CPAN. Use the "ppm query" command to check the exact
       version included in this release.

  Build 2200 June, 2015
    *What's new in the 2200 Series*

    *  This build corresponds to the Perl 5.22.0 source code release.

    *  The 2200 series builds of ActivePerl are not binary compatible with
       builds in the 2000 or earlier series. Any extensions built using
       binaries from the ActivePerl 2000 or earlier series will need to be
       recompiled. Note especially that this applies to PPM packages that
       may have been built for earlier series of ActivePerl.

    *  Significant changes that have occurred since the 5.20 release are
       documented in perl5220delta.

    *  Most bundled distributions have been updated to their latest
       released version from CPAN. Use the "ppm list" command to check the
       exact version included in this release.



perl/lib/metaclass.pm
package metaclass;
our $VERSION = '2.1603';

use strict;
use warnings;

use Module::Runtime 'use_package_optimistically', 'use_module';
use Class::MOP;

sub import {
    my ( $class, @args ) = @_;

    unshift @args, "metaclass" if @args % 2 == 1;
    my %options = @args;

    my $meta_name = exists $options{meta_name} ? $options{meta_name} : 'meta';
    my $metaclass = delete $options{metaclass};

    unless ( defined $metaclass ) {
        $metaclass = "Class::MOP::Class";
    } else {
        use_package_optimistically($metaclass);
    }

    ($metaclass->isa('Class::MOP::Class'))
        || die use_module('Moose::Exception::MetaclassMustBeDerivedFromClassMOPClass')->new( class_name => $metaclass );

    # make sure the custom metaclasses get loaded
    foreach my $key (grep { /_(?:meta)?class$/ } keys %options) {
        unless ( ref( my $class = $options{$key} ) ) {
            use_package_optimistically($class)
        }
    }

    my $package = caller();

    # create a meta object so we can install &meta
    my $meta = $metaclass->initialize($package => %options);
    $meta->_add_meta_method($meta_name)
        if defined $meta_name;
}

1;

# ABSTRACT: a pragma for installing and using Class::MOP metaclasses

__END__

=pod

=encoding UTF-8

=head1 NAME

metaclass - a pragma for installing and using Class::MOP metaclasses

=head1 VERSION

version 2.1603

=head1 SYNOPSIS

  package MyClass;

  # use Class::MOP::Class
  use metaclass;

  # ... or use a custom metaclass
  use metaclass 'MyMetaClass';

  # ... or use a custom metaclass
  # and custom attribute and method
  # metaclasses
  use metaclass 'MyMetaClass' => (
      'attribute_metaclass' => 'MyAttributeMetaClass',
      'method_metaclass'    => 'MyMethodMetaClass',
  );

  # ... or just specify custom attribute
  # and method classes, and Class::MOP::Class
  # is the assumed metaclass
  use metaclass (
      'attribute_metaclass' => 'MyAttributeMetaClass',
      'method_metaclass'    => 'MyMethodMetaClass',
  );

  # if we'd rather not install a 'meta' method, we can do this
  use metaclass meta_name => undef;
  # or if we'd like it to have a different name,
  use metaclass meta_name => 'my_meta';

=head1 DESCRIPTION

This is a pragma to make it easier to use a specific metaclass
and a set of custom attribute and method metaclasses. It also
installs a C<meta> method to your class as well, unless C<undef>
is passed to the C<meta_name> option.

Note that if you are using Moose, you most likely do B<not> want
to be using this - look into L<Moose::Util::MetaRole> instead.

=head1 AUTHORS

=over 4

=item *

Stevan Little <stevan.little@iinteractive.com>

=item *

Dave Rolsky <autarch@urth.org>

=item *

Jesse Luehrs <doy@tozt.net>

=item *

Shawn M Moore <code@sartak.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Karen Etheridge <ether@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

Hans Dieter Pearcey <hdp@weftsoar.net>

=item *

Chris Prather <chris@prather.org>

=item *

Matt S Trout <mst@shadowcat.co.uk>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Infinity Interactive, Inc..

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
